### PR TITLE
Updated UltraGrids video pipeline to use threading in two key areas

### PR DESCRIPTION
**Pull request by aw-sohonet**
- Sending packets
  - This causes the sending of packets using the RTP stack to be threaded amongst all available CPU threads.
  - This was implemented because sending uncompressed / lossless frames at 24 FPS was often taking longer to send the frame than than the duration of the frame. As the frames are collected in the same thread that they are sent in, this can cause a slowing in the collection of frames. Currently this option is forced. Future work will make this optional.
  - A huge increase in performance has been measured, such that lossless / uncompressed frames are able to be processed quickly enough on consumer hardware (that has enough thread capability).
  - A parameter has been introduced to control the amount of avaialble threads to be used for encoding ("encode-thread-count").
- Receiving packets
  - Due to the nature of the way packets are received in UltraGrid there has been two threading implementations placed into the decode pipeline.
  - The first of which is that decoding the packets into the initial package is now run in a thread independently of the build-up of packets. This helps avoid queues in the kernel for incoming UDP packets.
  - The second of which is a series of threads that are used to decrypt the contents of the packets and place them into the frame. These threads do not lock when writing into the frame (as each packet should represent an independent region of the memory).
  - A parameter has been introduced to control the amount of available thread to be used for decoding and decrypting packets ("decode-thread-count").

Some fixes for the Reed Solomon implementation for audio have also been introduced (mostly focused on a memory leak that existed in the implementation). A fix has also been implemented to remove the packet shaping from the audio FEC RTP send function that was causing a crash.

### CODE REVIEW

File: src/crypto/crypt_aes.c [hunk 1]
```diff
@@ -189,7 +189,8 @@ int blockEncrypt(cipherInstance * cipher, keyInstance * key,
         case MODE_CFB1:
                 iv = cipher->IV;
                 for (i = numBlocks; i > 0; i--) {
-                        memcpy(outBuffer, input, 16);
+                        
+memcpy(outBuffer, input, 16);
```
Comment (heindsight) [inline src/crypto/crypt_aes.c]: Indentation seems to have gone a bit wonky here.

File: src/crypto/crypt_aes.c [hunk 2]
```diff
@@ -330,7 +331,8 @@ int blockDecrypt(cipherInstance * cipher, keyInstance * key,
                 iv = cipher->IV;
                 for (i = numBlocks; i > 0; i--) {
                         memcpy(outBuffer, input, 16);
-                        for (k = 0; k < 128; k++) {
+                        
+for (k = 0; k < 128; k++) {
```
Comment (heindsight) [inline src/crypto/crypt_aes.c]: And here

File: src/rtp/rtp.cpp [hunk 3]
```diff
@@ -3977,7 +3984,8 @@ static bool rijndael_decrypt(struct rtp *session, unsigned char *data,
  */
 bool rtp_set_recv_buf(struct rtp *session, int bufsize)
 {
-        return udp_set_recv_buf(session->rtp_socket, bufsize);
+    return udp_set_recv_buf(session->rtp_socket, bufsize);
+    //return udp_set_recv_buf(session->rtp_socket, 100000000);
```
Comment (heindsight) [inline src/rtp/rtp.cpp]: Left-over debugging code?

File: src/crypto/crypt_aes.c (hunk 1 above)
Comment (aw-sohonet) [reply src/crypto/crypt_aes.c]: You're right! I'm not sure how that ended up happening - Fixed.

File: src/crypto/crypt_aes.c (hunk 2 above)
Comment (aw-sohonet) [reply src/crypto/crypt_aes.c]: Fixed!

File: src/rtp/rtp.cpp (hunk 3 above)
Comment (aw-sohonet) [reply src/rtp/rtp.cpp]: Good catch, thanks!

**Pull request merged**